velox/exec/Window.cpp

```diff
@@ -379,25 +451,56 @@ void Window::callApplyForPartitionRows(
         }
         break;
       }
-      case core::WindowNode::BoundType::kPreceding:
-      case core::WindowNode::BoundType::kFollowing:
-        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kPreceding: {
+        updateKRowsFrameBounds(
+            true, frameArg.value(), startRow, numRows, rawFrameBounds);
+        break;
+      }
+      case core::WindowNode::BoundType::kFollowing: {
+        updateKRowsFrameBounds(
+            false, frameArg.value(), startRow, numRows, rawFrameBounds);
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  // k preceding and k following bounds in ROWS mode can go over the partition
+  // limits. Hence they are bound to the first and last partition rows.
+  auto fixOutOfBoundFrames = [&](vector_size_t* frameStart,
```

Comment:
This can be moved to a simple method in an anonymous namespace.

Reply:
Updated, thanks for the suggestion!